File: cli/mosaic_cli/core/tools/run_command.py
import subprocess
from .base import Tool
from typing import Dict, Any

class RunCommandTool(Tool):
    def name(self) -> str:
        return "run_command"

    def description(self) -> str:
        return "Execute a shell command in the workspace."

    async def execute(self, params: Dict[str, Any], workspace: str) -> str:
        import re
        command = params.get("command", "")
        if not command:
            return "Error: Missing command parameter"
        
        # Safety check: Block commands that target protected hidden files
        from .utils import is_protected_path
        
        # Find potential hidden targets in the command
        hidden_targets = re.findall(r"(?:^|\s|/|'|\")(\.[a-zA-Z0-9_-]+)", command)
        for target in hidden_targets:
            # If the target is protected, check if the command is "destructive"
            if is_protected_path(target, workspace):
                destructive_op = re.search(r"\b(rm|mv|cp|tar|zip)\b|>", command)
                if destructive_op:
                    return f"Error: Command denied. Destructive operation targeting protected file/directory '{target}' is restricted."
        
        try:
            # Using shell=True can be dangerous, but it's often needed for complex commands
            result = subprocess.run(
                command,
                shell=True,
                cwd=workspace,
                capture_output=True,
                text=True,
                timeout=300
            )
            
            output = result.stdout
            if result.stderr:
                output += f"\n--- Error Output ---\n{result.stderr}"
            
            if result.returncode != 0:
                output += f"\n--- Return Code: {result.returncode} ---"
                
            return output if output.strip() else "(command completed with no output)"
        except subprocess.TimeoutExpired:
            return "Error: Command timed out after 5 minutes"
        except Exception as e:
            return f"Error: {str(e)}"

File: vscode/src/core/tools/runCommand.ts
import * as vscode from 'vscode';
import * as path from 'path';
import * as cp from 'child_process';
import { BaseTool } from './base';

export class RunCommandTool extends BaseTool {
  name() { return "run_command"; }
  description() { return "Run a shell command in the integrated VSCode terminal. Use this for builds, installs, git, and testing."; }
  schema() {
    return {
      type: "object",
      properties: {
        command: { type: "string", description: "The shell command to execute" }
      },
      required: ["command"]
    };
  }

  async execute(args: { command: string }) {
    if (!args.command) return this.formatError("No command provided");
    
    const safetyError = this.validateCommand(args.command);
    if (safetyError) {
      return this.formatError(`SAFETY BLOCK: ${safetyError}`);
    }

    const workspaceFolders = vscode.workspace.workspaceFolders;
    const cwd = workspaceFolders ? workspaceFolders[0].uri.fsPath : process.cwd();

    // Show in terminal for visual feedback
    let terminal = vscode.window.terminals.find(t => t.name === "Mosaic Terminal");
    if (!terminal) {
      terminal = vscode.window.createTerminal("Mosaic Terminal");
    }
    terminal.show(true);
    terminal.sendText(args.command);

    // Execute and capture output
    return new Promise((resolve) => {
      cp.exec(args.command, { cwd, timeout: 60000, maxBuffer: 1024 * 1024 }, (error: any, stdout: string, stderr: string) => {
        if (error) {
          resolve(this.formatError(`Command failed with exit code ${error.code}\nStdout: ${stdout}\nStderr: ${stderr}`));
          return;
        }
        resolve(stdout || stderr || "Command executed successfully (no output)");
      });
    });
  }

  private validateCommand(command: string): string | null {
    const workspaceFolders = vscode.workspace.workspaceFolders;
    const rootPath = workspaceFolders ? workspaceFolders[0].uri.fsPath : null;

    const dangerousPatterns = [
      { pattern: new RegExp("rm\\s+-rf\\s+[\\/~]"), reason: "Destructive operations on root (/) or home (~) directories are strictly forbidden." },
      { pattern: new RegExp("rm\\s+-rf\\s+\\.(git|mosaic)"), reason: "Critical project directories (.git, .mosaic) must be preserved." },
      { pattern: new RegExp("mkfs|fdisk|parted"), reason: "Disk partitioning and formatting commands are blocked." },
      { pattern: new RegExp("dd\\s+.*of=\\/dev\\/"), reason: "Direct block device writes via 'dd' are strictly forbidden." },
      { pattern: new RegExp(":\\(\\)\\{\\s*:\\s*\\|\\s*:\\s*&\\s*\\}\\s*;\\s*:"), reason: "Fork bomb detected." },
      { pattern: new RegExp("shutdown|reboot|halt"), reason: "System power commands are blocked." }
    ];

    const normalized = command.toLowerCase().trim();
    for (const { pattern, reason } of dangerousPatterns) {
      if (pattern.test(normalized)) return reason;
    }

    // Home directory protection
    if (command.includes('~')) {
      return "Access to home directory (~) is strictly forbidden.";
    }

    // Path traversal protection
    if (command.includes('..')) {
      const upCount = (command.match(/\.\.\//g) || []).length;
      if (upCount > 0) {
        // Even one '..' can be dangerous if it goes above root.
        // For simplicity and maximum safety, we block '..' in commands unless strictly needed.
        // But some build commands might use them. Let's be conservative.
        if (upCount > 2) return "Potential path traversal detected.";
      }
    }

    // Absolute path protection
    if (rootPath) {
      // Improved regex to find absolute paths anywhere in the command
      const absPathRegex = /(?:^|\s)(\/[^\s;&|]+)/g;
      let match;
      while ((match = absPathRegex.exec(command)) !== null) {
        const p = match[1].trim();
        if (path.isAbsolute(p) && !p.startsWith(rootPath)) {
           return `Unauthorized access to absolute path outside repository: ${p}`;
        }
      }
    }

    return null;
  }
}
